src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/WhyCommand/WhyCommandRunner.cs

```diff
@@ -0,0 +1,548 @@
+        private List<DependencyNode> GetDependencyGraphPerFramework(
+            IEnumerable<InstalledPackageReference> topLevelPackages,
+            IList<LockFileTargetLibrary> packageLibraries,
+            string targetPackage)
+        {
+            List<DependencyNode> dependencyGraph = null;
+
+            // hashset tracking every package node that we've traversed
+            var visited = new HashSet<string>();
+            // dictionary tracking all package nodes that have been added to the graph, mapped to their DependencyNode objects
+            var dependencyNodes = new Dictionary<string, DependencyNode>();
+            // dictionary mapping all packageIds to their resolved version
+            Dictionary<string, string> versions = GetAllResolvedVersions(packageLibraries);
+
+            foreach (var topLevelPackage in topLevelPackages)
+            {
+                // use depth-first search to find dependency paths from the top-level package to the target package
+                DependencyNode topLevelNode = FindDependencyPath(topLevelPackage.Name, packageLibraries, visited, dependencyNodes, versions, targetPackage);
+
+                if (topLevelNode != null)
+                {
+                    dependencyGraph ??= [];
+                    dependencyGraph.Add(topLevelNode);
+                }
+            }
+
+            return dependencyGraph;
+        }
+
+
+        /// <summary>
+        /// Adds all resolved versions of packages to a dictionary.
+        /// </summary>
+        /// <param name="packageLibraries">All package libraries for a given framework.</param>
+        /// <returns></returns>
+        private Dictionary<string, string> GetAllResolvedVersions(IList<LockFileTargetLibrary> packageLibraries)
+        {
+            var versions = new Dictionary<string, string>();
+
+            foreach (var package in packageLibraries)
+            {
+                versions.Add(package.Name, package.Version.ToNormalizedString());
+            }
+
+            return versions;
+        }
+
+        /// <summary>
+        /// Traverses the dependency graph for a given top-level package, looking for a path to the target package.
+        /// Returns the top-level package node if a path was found, and null if no path was found.
+        /// </summary>
+        /// <param name="topLevelPackage">Top-level package.</param>
+        /// <param name="packageLibraries">All package libraries for a given framework.</param>
+        /// <param name="visited">HashSet tracking every package node that we've traversed.</param>
+        /// <param name="dependencyNodes">Dictionary tracking all packageIds that were added to the graph, mapped to their DependencyNode objects.</param>
+        /// <param name="versions">Dictionary mapping packageIds to their resolved versions.</param>
+        /// <param name="targetPackage">The package we want the dependency paths for.</param>
+        /// <returns></returns>
```

Comment:
documentation: The guidance tends to not leave empty tags. I'd just document this here. 

Apply this to any other places in this changeset.